New data → color pipeline

trygvrad opened the pull request:
The `Colorizer` class, `ColorizerShim`, and `ColorableArtist` which replaces the old `ScalarMappable`.


## PR summary
This PR implements the changes in #28428 where a container class is used to hold the $data\rightarrow color$ pipeline. 
Schematically, this pipeline looks like this:
![image](https://github.com/user-attachments/assets/c6c902ca-d093-4581-98e5-1e9ff7852734)
I have changed the names to be:
```
Mapper → Colorizer
ScalarMappableShim → ColorizerShim
ColorableArtist → ColorizingArtist
```
compared to the existing pipeline:
![image](https://github.com/user-attachments/assets/d372d4d1-99d5-4eee-b76c-7e9025fdb35b)

This is a minimal implementation so that the interested parties can take a look.
It does not include typing or new tests, but all existing tests pass

---
## Note 1:

`changed()` on `Colorizer` and `ColorableArtist` is executes:
```py
        self.callbacks.process('changed')
```
This is consistent with how the callbacks are handled on the norm
but different from how the callbacks are handled on `ScalarMappable`
```py
        self.callbacks.process('changed', self)
```
I believe this it is more consistent if the new classes do `self.callbacks.process('changed')` and I have not found any example where the additional argument matters.

---
## Note 2:

I would like to move `Colorizer` and `ColorizerShim` to a new file, instead of `cm.py`.
If this is done, `cm.py` would only contain `ColormapRegistry` (and `ScalarMappable` for compatibility).

----

## Note 3:
Here are some points that could be addressed in following PRs:

1. Colorbar does not need a reference to the artist, only the Colorizer. A small PR could change this, but there should probably be an example on how you can change the `.colorizer` on both an artist and a colorbar.
2. With this PR, the following example can be greatly simplified, and should be updated: https://matplotlib.org/stable/gallery/images_contours_and_fields/multi_image.html
3. The  `ColorableArtist` class has a `cbook.CallbackRegistry()`, but I believe this to be superfluous as all callbacks should in reality come from the `Colorizer`.  
4. The bulk of `Artist.format_cursor_data` should be moved to `ColorizableArtist`.
5. Fix the bug in updating hexbin #28633
6. There are still some scattered uses of `cm.ScalarMappable`, where artists need additional ones. These should be changed to use the new `Colorizer`.
7. From @timhoffm : We may have to reconsider how auto-scaling is executed. Currently, the first array to be set triggers auto-scaling. This does not work well when multiple ColorizingArtists share a Colorizer and thus a Norm. It may be reasonable to defer scaling to the first time that the norm (or the colorizing pipeline) is executed and collect the data limits from all participating Artists at that time. However, this requires a connection from the Colorizer to all associated artists. - I suppose this should be handled on the Colorizer level, The norm itself should likely stay Artist-agnostic.
---
# Note 4: 

The following code demonstrates how multiple axes can be connected to one `Colorizer` object:
```py
import matplotlib
matplotlib.use('qtagg')
import matplotlib.pyplot as plt
import numpy as np

np.random.seed(19680801)
data = np.linspace(0.1, 0.6, 6)[:, np.newaxis, np.newaxis] * np.random.rand(6, 10, 20)

colrz = matplotlib.cm.Colorizer()
colrz.vmin = np.min(data)
colrz.vmax = np.max(data)

fig, axs = plt.subplots(2, 3)
fig.suptitle('Multiple images')
for i, ax in enumerate(axs.ravel()):
    im = ax.imshow(data[i], norm = colrz)
    ax.label_outer()

cb = fig.colorbar(im, ax=axs, orientation='horizontal', fraction=.1)

plt.show()
```
(I didn't re-record the video, it was orignally posted [here](https://github.com/matplotlib/matplotlib/pull/28428#issuecomment-2225849569) and in order for changes made on the colorbar to propagate to the artist the `Colorbar` class needs some small changes [see point 1. in the list above], but changes on one of the images will propagate to all images and the colorbar as shown in the video.)
https://github.com/user-attachments/assets/093d5fa1-0dfc-4a68-9197-24acd0ef4aed


## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [N/A] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [¹] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

¹ we should consider the existing tests for `cm.ScalarMappable` and see if any of them should be replicated to test `Colorizer` and `ColorableArtist`.

timhoffm commented:
Naming: Not sure `ScalarMappableShim → ColorizerShim` is meaningful. I would say, this a *Shim for the ScalarMappable API* and hence, `ScalarMappable` in the name makes sense.

Note 2: Yes, please move to a new file.

Note 3: All agreed. Note that I've already simplified https://matplotlib.org/devdocs/gallery/images_contours_and_fields/multi_image.html through #28545. This still needs adaption to the Colorizer API, but only after we've untangled the Artist backreferences so that multiple Artists can share the same Colorizer/Colorbar

timhoffm commented on lib/matplotlib/artist.py:
```diff
@@ -1417,6 +1418,80 @@ def set_mouseover(self, mouseover):
+                            "converted to float")
+
+        self._A = A
+        if not self.norm.scaled():
```
Side-note: We may have to reconsider how auto-scaling is executed. Currently, the first array to be set triggers auto-scaling. This does not work well when multiple ColorizingArtists share a Colorizer and thus a Norm. It may be reasonable to defer scaling to the first time that the norm (or the colorizing pipeline) is executed and collect the data limits from all participating Artists at that time. However, this requires a connection from the Colorizer to all associated artists. - I suppose this should be handled on the Colorizer level, The norm itself should likely stay Artist-agnostic.

This should be considered / changed as part of the Artist connection refactoring (Note 3.1)

trygvrad replied:
I added this to the list above

timhoffm replied:
Marking as resolved because this is tracked above.

trygvrad commented:
@timhoffm 
> Naming: Not sure ScalarMappableShim → ColorizerShim is meaningful. I would say, this a Shim for the ScalarMappable API and hence, ScalarMappable in the name makes sense.

My thoughts regarding the name have evolved as I have been working on the issue, which has led me to the following points:
1. The name `ScalarMappableShim` makes sense in the context that we are replacing `ScalarMappable`, but in a few years time, with `ScalarMappable` possibly depreciated and new developers coming in, I don't think the name `ScalarMappableShim` will make much sense any more.
2. The shim allows access to attributes on a `Colorizer` objects without referencing the object itself¹, and this functionality is useful for the end user, i.e. while we could depreciate `ScalarMappable` in the foreseeable future, I don't really see a future where we depreciate the shim.
3. The shim will support both scalar and multivariate data, and for that reason I would like to avoid 'Scalar' in the name.

That said, I don't think the name is of much significance, so just tell me again and I'll change it back to `ScalarMappableShim` :slightly_smiling_face: 

¹ In my mind `ColorizerShim` well describes this use, but I may be misusing 'shim' as a technical term, as I am not familiar with the terminology.

trygvrad commented:
@timhoffm 
I moved `Colorizer` to a new file (this also makes the git diff a lot easier to read)

Also: 
This still needs updated typing and tests, but I would like to holding off on spending too much time on this until I get a bit more feedback @tacaswell @ksunden @story645.

timhoffm commented on lib/matplotlib/artist.py:
```diff
@@ -1346,7 +1347,8 @@ def format_cursor_data(self, data):
         get_cursor_data
         """
-        if np.ndim(data) == 0 and isinstance(self, ScalarMappable):
+        if np.ndim(data) == 0 and (isinstance(self, ScalarMappable) or
+                                   isinstance(self, ColorizingArtist)):
```
```suggestion
        if np.ndim(data) == 0 and isinstance(self, (ScalarMappable, ColorizingArtist)):
```

timhoffm commented on lib/matplotlib/artist.py:
```diff
@@ -1417,6 +1419,80 @@ def set_mouseover(self, mouseover):
+            interval ``[0, 1]``.
+            If a `str`, a `colors.Normalize` subclass is dynamically generated based
+            on the scale with the corresponding name.
+            If `cm.Colorizer`, the norm an colormap on the `cm.Colorizer` will be used
```
I would not overload the signature. A typical strategy would be to have the canoncial signature and create a factory classmethod for the other signature:
```
def __init__(colorizer):
    self.colorizer = colorizer

@classmethod
def from_norm_and_cmap(cls, norm, cmap):
    return cls(Colorizer(norm, cmap)
```
AFAICT, the situation here is a bit different, because users do not instantiate `ColorizingArtist` directly, and hence, they won't use the factory method. We could do one of the two options:

- only implement `def __init__(colorizer)`. If subclasses need to initialize from norm + cmap, they'd simply do
  ```
  ColorizingArtist.__init__(Colorizer(norm, cmap))
  ```
- create an alternative initializer
  ```
  def _init_from_norm_and_cmap(norm, cmap):
      self.colorizer = Colorizer(norm, cmap)
```
given that we'll only have a handfull of subclasses and the solution is quite simple, I'm leaning towards the first option.

trygvrad replied:
The first alternative makes sense to me.

However, we still need the logic where the user can pass a `Colorizer` object as part of the input to the plotting function.

I implemented this with the following function:
```py
def _get_colorizer(cmap, norm):
    """
    Passes or creates a Colorizer object.

    Allows users to pass a Colorizer as the norm keyword
    where a artist.ColorizingArtist is used as the artist.
    If a Colorizer object is not passed, a Colorizer is created.
    """
    if isinstance(norm, Colorizer):
        if cmap:
            raise ValueError("Providing a `cm.Colorizer` as the norm while "
                             "at the same time providing a `cmap` is not supported.")
        return norm
    return Colorizer(cmap, norm)
```
Such that the initiation of a `ColorizingArtist` in e.g. `Collection` is:
```py
        artist.ColorizingArtist.__init__(self, colorizer._get_colorizer(cmap, norm))
```

timhoffm replied:
Do we really want to pass a `Colorizer` via the `norm` parameter in public API? This feels very hacky. I'm inclined to add a `colorizer` parameter instead, even if that means we have mutually exclusive parameters. (Note that even with the hack, we'd have to live with partially mutually exclusive parameters, because you cannot pass `norm=mycolorizer, cmap='jet'`).

trygvrad replied:
I have no strong opinions, and am inclined to agree with you @timhoffm.
Since this is a change to the call signature of the plotting functions, we should bring it up at a weekly meeting. @story645 could you add it on the agenda for the next meeting?

(My aversion to new keyword arguments is a result of a discussion at a weekly meeting a long time ago about multivariate color mapping, and how we do not want *that* to increase the number of keyword arguments, but this is a different case entirely.)

timhoffm commented on lib/matplotlib/collections.py:
```diff
@@ -32,7 +32,7 @@
     "offset_transform": ["transOffset"],
 })
-class Collection(artist.Artist, cm.ScalarMappable):
+class Collection(artist.ColorizingArtist, colorizer.ColorizerShim):
```
No. The shim provides the backward-compatible interface for the ScalarMappaple API. It‘s needed for the foreseeable future, but it’s reasonable to keep it separated from the ColorizingArtist API.

timhoffm commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,483 @@
+        if norm is None:
+            norm = colors.Normalize()
+        elif isinstance(norm, str):
+            try:
+                scale_cls = scale._scale_mapping[norm]
+            except KeyError:
+                raise ValueError(
+                    "Invalid norm str name; the following values are "
+                    f"supported: {', '.join(scale._scale_mapping)}"
+                ) from None
```
Rather https://matplotlib.org/devdocs/api/_api_api.html#matplotlib._api.check_getitem

trygvrad replied:
This code mirrors the existing code for ScalarMappable (cm.py lines 545→552), but I'll see if I can shift it to use the _api or if there is some reason why that has not been implemented for ScalarMappable.

trygvrad replied:
Two points:
1. This will change the error message from:
```
" Invalid norm str name; the following values are supported: 'linear', 'log', 'symlog', 'asinh', 'logit', 'function', 'functionlog'
``` 
to
```
' ' is not a valid value for norm; supported values are 'linear', 'log', 'symlog', 'asinh', 'logit', 'function', 'functionlog'
``` 
Which is slightly misleading, as norm supports other inputs than strings. We can mitigate this by intercepting the error message and replacing **value** with **string**, but this requires an additional try except block.

2. I think this check naturally belongs in `_auto_norm_from_scale`, a function which is currently called by:
```py
        elif isinstance(norm, str):
            try:
                scale_cls = scale._scale_mapping[norm]
            except KeyError:
                raise ValueError(
                    "Invalid norm str name; the following values are "
                    f"supported: {', '.join(scale._scale_mapping)}"
                ) from None
            norm = _auto_norm_from_scale(scale_cls)()
```
which I think should be replaced by:
```py
        elif isinstance(norm, str):
            norm = _auto_norm_from_str(norm)()
```
(note the change in function name from `_scale` to `_str`)
This function then incorporates the check.
```py
def _auto_norm_from_str(norm_str):
    scale_cls = _api.check_getitem(scale._scale_mapping, norm=norm_str)
    ...
```
@timhoffm @story645 Let me know your opinions on point 1 and 2 above, and I will implement it in this PR.
Point 2. would be a useful change with regards to making a 'MultiNorm' in a later PR.

timhoffm replied:
I agree the message would be slightly misleading. OTOH pushing this down does not make much better for the user. Thy still get the same exception message. Only the call stack is different. While the message in the deeper context is correct, it's further away from the public API and thus more difficult to trace back.

An alternative would be expanding `_check_get_item` so that we can do `_api.check_getitem(scale._scale_mapping, norm=norm_str, _type="str")` and get "' ' is not a valid str value ...", which would imply there may be other valid types.

I believe in practice, it does not matter. If the user has passed an invalid str, they most likely have a typo and still wanted to pass a str name. It's unlikely that the information being able to pass Norms directly would be helpful. To keep things simple, I slighly favor just using `_check_get_item` as is in original context. But no strong opinion.

trygvrad replied:
I think it's better to have it here, with the changes in #28690 that means we only have to move this function once.

trygvrad commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,483 @@
+        performed, and it is assumed to be in the range (0-1).
+
+        """
+        # First check for special case, image input:
+        # First check for special case, image input:
```
This version is how to_rgba() has been implemented in ScalarMappable, and in the interest of minimal changes I have kept this implementation for Colorizer. 
It would of course be more elegant to factor it out.

trygvrad replied:
I factored out the pipeline for data that is already RGB(A), i.e. arrays of shape (n, m, 3(4)) to a new static function `_pass_image_data`, called as follows:
```py
        # First check for special case, image input:
        if isinstance(x, np.ndarray) and x.ndim == 3:
            return self._pass_image_data(x, alpha, bytes, norm)
```
---
Previously, there was a try-except block here, which I replaced with a type check. 
i.e. previously there was:
```py
        # First check for special case, image input:
        try:
            if x.ndim == 3:
                ...
        except AttributeError:
            # e.g., x is not an ndarray; so try mapping it
            pass
```
This should have no implications for the end user, as all input data passes through `ColorizingArtist.set_data` (previously `ScalarMappable.set_data`) which calls `cbook.safe_masked_invalid` which converts all input data to np.ndarray or np.ma.MaskedArray.

trygvrad commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,483 @@
+class Colorizer():
+    """
+    Class that holds the data to color pipeline
+    accessible via `.to_rgba(A)` and executed via
```
Seeing as this is the main operation of the `Colorizer` class, I changed the the name of this function from `to_rgba()` to a simple `__call__`. 
@timhoffm do you agree that `Colorizer.__call__` is the best call signature for the norm→color pipeline?

timhoffm replied:
I'm undecided. My general point of view is that there are only very few cases where `__call__` is justified (mostly drop-in replacements for callables). It's generally better to have an explicit name for the action; otherwise, you end up with stuff like `plt.get_cmap('jet')(0.5)`. OTOH norms and cmaps work that way, so it's not completely unreasonable to keep that pattern for `Colorizer`.

trygvrad replied:
I changed it back to to_rgba, se Toms comment below.

trygvrad commented on lib/matplotlib/artist.py:
```diff
@@ -1417,6 +1419,67 @@ def set_mouseover(self, mouseover):
+        A : array-like or None
+            The values that are mapped to colors.
+
+            The base class `.VectorMappable` does not make any assumptions on
+            the dimensionality and shape of the value array *A*.
```
yep, fixed

trygvrad commented on lib/matplotlib/cm.py:
```diff
@@ -277,40 +275,13 @@ def get_cmap(name=None, lut=None):
+    The VectorMappable applies data normalization before returning RGBA colors
+    from the given `~matplotlib.colors.Colormap`, `~matplotlib.colors.BivarColormap`,
+    or `~matplotlib.colors.MultivarColormap`.
     """
```
that is a holdover from a previous version, fixed.

tacaswell commented on lib/matplotlib/artist.py:
```diff
@@ -1417,6 +1419,67 @@ def set_mouseover(self, mouseover):
+                            "converted to float")
+
+        self._A = A
+        if not self.norm.scaled():
```
are we sure that this will always have a `.norm`?

trygvrad replied:
I'm thinking it will always have a norm, as I think that simplifies the implementation. 
However, that norm might be `NoNorm`.

The implementation here mirrors that of `ScalarMappable`, which always has a `Normalize` and `Colormap` object, and I think this is sensible. For images that are already RGB(A) a norm (and cmap) is technically unnecessary, but I believe the overhead of creating them is negligible.

tacaswell commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,486 @@
+        if not in_init:
+            self.changed()
+
+    def __call__(self, x, alpha=None, bytes=False, norm=True):
```
Given that we are almost always going to use this in library code (and only type it out in a terminal rarely) I am weakly against using `__call__`.  That we can do this is Python is sometimes very nice from an API point of view, but it is mostly in the case where you want to hide the fact that you up-converted from a plain function (with usually no state) to an object that holds a bunch of state (or is acting as a "mutable closure").

Even though we use this pattern with norm and colormap, I think we should break it here and give this a name!

timhoffm replied:
:smile: Almost the same points I made [above](https://github.com/matplotlib/matplotlib/pull/28658#discussion_r1710353662). Let's give it a name.

trygvrad replied:
I changed it back to `to_rgba`, since that is already in use in the library, I think that makes the most sense

tacaswell commented:
> The shim provides the backward-compatible interface for the ScalarMappaple API. It‘s needed for the foreseeable future, but it’s reasonable to keep it separated from the ColorizingArtist API.

The norm and cmap properties only ever make sense with the `ColorizerArist` as the mix-in assumes that something else is providing the `.colorizer` attribute.  I'm very unconvinced that splitting it off into its own class does more than make our multiple inheritance look worse and add some extra ceremony.  If we do want to have a version of `CololizerArtist` without the back-compat API can we call it `ColorizerCore` or something like that and have a sub-class (that we use everywhere) that include the properties?

timhoffm commented:
> > The shim provides the backward-compatible interface for the ScalarMappaple API. It‘s needed for the foreseeable future, but it’s reasonable to keep it separated from the ColorizingArtist API.
> 
> The norm and cmap properties only ever make sense with the `ColorizerArist` as the mix-in assumes that something else is providing the `.colorizer` attribute. I'm very unconvinced that splitting it off into its own class does more than make our multiple inheritance look worse and add some extra ceremony. If we do want to have a version of `CololizerArtist` without the back-compat API can we call it `ColorizerCore` or something like that and have a sub-class (that we use everywhere) that include the properties?

My motivation was to have a separation between "API we have to provide to ensure backward compatibility" and "API we want to provide". At least during the design phase this could make it more clear. We also need to decide at some point whether there is API that we want to discourage. This is all possible also with a single large class, but I have the impression that that is more mental load and we have to be extra careful to not miss anything. But no strong opinion. If you think you can  keep the overview, that's ok.

trygvrad commented:
I'm going to make `_ColorizingInterface` as a base class which contains the functions that are now part of the shim. 
Then I will make `ColorizingArtist` inherit from `_ColorizingInterface` and `Artist` 
This has two advantages:
1. It separates the functions that do things on the `ColorizingArtist` from the functions that only delegate to the `Colorizer`
2. We can let `ScalarMappable` inherit from `_ColorizingInterface` so that we don't need to keep two versions of these functions around (for as long as `ScalarMappable` is around).

Once/if `ScalarMappable` is deprecated we can evaluate if it makes sense to remove `_ColorizingInterface` as a separate class.

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -4696,7 +4697,7 @@ def invalid_shape_exception(csize, xsize):
     @_docstring.interpd
     def scatter(self, x, y, s=None, c=None, marker=None, cmap=None, norm=None,
-                vmin=None, vmax=None, alpha=None, linewidths=None, *,
+                vmin=None, vmax=None, colorizer=None, alpha=None, linewidths=None, *,
```
This has to go after the `*` or we break API (even though passing 10 positional arguments is a Bad Idea ™️ if someone was doing that we would now get the value of `alpha` as `colorizer`).

timhoffm replied:
Side-remark: We should consider tightening on the kw-only settings. It helps the understanding to collect related parameters next to each other. I'd be willing to force people with Bad Ideas™️ to adjust in this case.

But this doesn't help for now because we'd need to go through the deprecation cycle. So @tacaswell's comment is valid anyway.

trygvrad replied:
Thank you for the feedback, I was unsure if it was OK to break the list of args for people with Bad Ideas™️, but I guess we should cater to everyone :)

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -4975,7 +4985,7 @@ def scatter(self, x, y, s=None, c=None, marker=None, cmap=None, norm=None,
     def hexbin(self, x, y, C=None, gridsize=100, bins=None,
                xscale='linear', yscale='linear', extent=None,
-               cmap=None, norm=None, vmin=None, vmax=None,
+               cmap=None, norm=None, vmin=None, vmax=None, colorizer=None,
```
same as above, this needs to go last.

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5728,7 +5745,7 @@ def fill_betweenx(self, y, x1, x2=0, where=None,
     def imshow(self, X, cmap=None, norm=None, *, aspect=None,
                interpolation=None, alpha=None,
-               vmin=None, vmax=None, origin=None, extent=None,
+               vmin=None, vmax=None, colorizer=None, origin=None, extent=None,
```
this one is OK because we made a bunch of these keyword only already!

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -6082,7 +6104,7 @@ def _interp_grid(X):
     @_docstring.dedent_interpd
     def pcolor(self, *args, shading=None, alpha=None, norm=None, cmap=None,
-               vmin=None, vmax=None, **kwargs):
+               vmin=None, vmax=None, colorizer=None, **kwargs):
```
this one is OK because `*args` means none of these can be passed positionally.

tacaswell commented on lib/matplotlib/image.py:
```diff
@@ -258,8 +260,8 @@ def __init__(self, ax,
                  ):
-        martist.Artist.__init__(self)
-        cm.ScalarMappable.__init__(self, norm, cmap)
+        mcolorizer.ColorizingArtist.__init__(self, mcolorizer._get_colorizer(cmap, norm,
```
```suggestion
        super().__init__(self, mcolorizer._get_colorizer(cmap, norm,
```

tacaswell commented on lib/matplotlib/image.py:
```diff
@@ -331,7 +333,7 @@ def changed(self):
         """
         self._imcache = None
-        cm.ScalarMappable.changed(self)
+        mcolorizer.ColorizingArtist.changed(self)
```
```suggestion
        super().changed(self)
```

tacaswell commented on lib/matplotlib/image.py:
```diff
@@ -1351,7 +1358,7 @@ def make_image(self, renderer, magnification=1.0, unsampled=False):
     def set_data(self, A):
         """Set the image array."""
-        cm.ScalarMappable.set_array(self, A)
+        mcolorizer.ColorizingArtist.set_array(self, A)
```
```suggestion
        super().set_array(self, A)
```

tacaswell commented:
I think this is heading the right direction.  The API issues need to be addressed, take or leave the `super()` comments, they are good to do but not critical or worth slowing down this PR over.

trygvrad commented:
I moved the `colorizer` keyword so that it always follows * or *args, and used `super()` instead of directly invoking the parent method.
@story645 @ksunden could you take a look at this PR now?

timhoffm commented:
> but one or both should confirm

I'm on mobile-only for the next week. Please do not expect a detailed review from my side in that time. Last time I checked, this looked pretty good. Feel free to move without me if you want to go ahead. I will re-check afterwards, but I don't expect any more serious issues.

trygvrad commented on lib/matplotlib/colorbar.py:
```diff
@@ -490,12 +490,13 @@ def update_normal(self, mappable):
-        self.cmap = mappable.cmap
-        if mappable.norm != self.norm:
-            self.norm = mappable.norm
+        if mappable:
+            self.mappable = mappable
```
`self.mappable` should never be None, as a `cm.ScalarMappable` is created if `mappable=None` on the creation of the colorbar. 

These code lines remain because they are in the existing codebase, and I don't want to change more than necessary.

Note that the colorbar currently connects to an `ColorizingArtist` (or a `ScalarMappable`). It would be more logically consistent to connect it to a `Colorizer` object, and this should be changed in a later PR.

Also, there is a difference in how `ScalarMappable` and `ColorizingArtist` (and `Colorizer`) emits changed signals. 
I added the following note:
```py
def update_normal(self, mappable=None):
    if mappable:
        # The mappable keyword argument exists because 
        # cm.ScalarMappable.changed() emits self.callbacks.process('changed', self)
        # in contrast, ColorizingArtist (and Colorizer) does not use this keyword.
        # [colorizer.ColorizingArtist.changed() emits self.callbacks.process('changed')]
        # Also, there is no test where self.mappable == mappable is not True and possibly no use case. 
        # Therefore, the mappable keyword can be depreciated if cm.ScalarMappable is removed.
        self.mappable = mappable
  ```

trygvrad commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5303,9 +5315,14 @@ def reduce_C_function(C: array) -> float
         collection.set_array(accum)
-        collection.set_cmap(cmap)
-        collection.set_norm(norm)
         collection.set_alpha(alpha)
         collection._internal_update(kwargs)
         collection._scale_norm(norm, vmin, vmax)
```
In my opinion the alpha keyword should always refer to the artist, because a primary use of the alpha keyword is to layer different artists in the same subplot, and you don't want the alpha to show up in the colorbar. With this logic, the `Colorizer` object should **not** have an alpha property, while `Artist` objects **should** have an alpha property (`Colorizer.to_rgba()` takes alpha as a keyword argument passed by the artist). 

I'm pretty sure this is how it is handled in the code as well.

In this specific case, `collections.Collection.set_alpha(alpha)` invokes `artist.Artist._set_alpha_for_array(alpha)`, which is independent of the `colorizer` framework.

ksunden commented on doc/api/colorizer_api.rst:
```diff
@@ -0,0 +1,9 @@
+*****************
+``matplotlib.colorizer``
+*****************
```
```suggestion
************************
``matplotlib.colorizer``
************************
```

Matching the over/underline length. Technically, it works, but

ksunden commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,631 @@
+
+.. admonition:: Provisional status of colorizer
+
+    The ``colorizer`` module and classes in this file are considered provisional and may change
```
This line is failing lint check (too long), just needs to be reflowed.

ksunden commented on doc/api/next_api_changes/behavior/28658-TMR.rst:
```diff
@@ -0,0 +1,7 @@
+ScalarMappable changed to ColorizingArtist
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Classes that used to inherit from `cm.ScalarMappable` now inherit from 
```
This file (on a couple lines) has trailing whitespace.

ksunden reviewed (commented):
This is mostly quick things that linters found, will do a more thorough review as well

tacaswell commented:
https://github.com/trygvrad/matplotlib/pull/1 is my proposal of how to do this without having to break `isinstanec(..., cm.ScalarMappable)`.  Breaking that with no warning seems like a pretty big user visible change.

trygvrad commented:
!!!! I forgot to run `pre-commit run --all-files`, my apologies !!!!

@story645 I read it the same way
@tacaswell that looks reasonable, lets talk about it at the meeting today.
Is the intent then to have a depreciation cycle for `ScalarMappable`?
I'm thinking it should have a note about how this can be done, i.e. 
```py


   # _ScalarMappable exists for comatibility with
   # code written before the introduction of the Colorizer
   # and ColorizingArtist classes.

   # _ScalarMappable can be depreciated so that ColorizingArtist
   # inherits directly from _ColorizerInterface.
   # in this case, the following changes should occur:
   # __init__() has its functionality moved to ColorizingArtist.
   # set_array(), get_array(), _get_colorizer() and 
   # _check_exclusionary_keywords() are moved to ColorizingArtist.
   # changed() can be removed so long as colorbar.Colorbar
   # is changed to connect to the colorizer instead of the 
   # ScalarMappable/ColorizingArtist,
   # otherwise changed() can be moved to ColorizingArtist.
```

---

The Colorizer docs now look like this:

https://trygvrad.github.io//mpl_docs/matplotlib.colorizer.html

I'm not super happy about how `_ColorizerInterface` is shown. I'd rather just list the member functions that are relevant for `ColorizingArtist` under `ColorizingArtist`, but I didn't find a way to do that without also listing the member functions of `Artist`. This issue gets compounded with the introduction of `_ScalarMappable`. 

If anyone has any tips in this regard, do let me know. 

I think the current version is *fine*, but once this PR is accepted I'll take another crack at it.

trygvrad commented:
@tacaswell @story645 I merged @tacaswell s PR, and added a note on how _ScalarMappalble can be deprecated.
It would be great if you could approve this now.

PR cleanliness fails because doc/api/next_api_changes/behaviour/28658-TMR.rts is both created and deleted, but this should be fixed with a squash merge.

timhoffm commented:
> but I didn't find a way to do that without also listing the member functions of Artist.

I believe this is a limitation of the sphinx autodoc. We have similar problems in other cases. The correct solution would be to fix it on the sphinx autodoc side and make the `:inherited-members:` configurable to only include given parent classes.

For now, there's no simple solution.

trygvrad commented:
@tacaswell @timhoffm would one of you like to approve this PR?
It is easier if this is approved before I make the next one (exposing multivariate color mapping to the plotting funcitons).

QuLogic commented:
The doc build should be fixed; a rebase at least should fix some of them.

QuLogic commented on lib/matplotlib/axes/_axes.pyi:
```diff
@@ -444,6 +446,7 @@ class Axes(_AxesBase):
         mincnt: int | None = ...,
         marginals: bool = ...,
         data=...,
+        colorizer: Colorizer | None = ...,
```
This should insert before `data`.

QuLogic commented on lib/matplotlib/collections.pyi:
```diff
@@ -4,9 +4,10 @@ from typing import Literal
+from . import colorizer, transforms
 from .backend_bases import MouseEvent
 from .artist import Artist
+from .colorizer import Colorizer
```
This seems redundant to `colorizer` being imported above. Probably only need one or the other way.

QuLogic commented on lib/matplotlib/colorbar.py:
```diff
@@ -490,12 +490,21 @@ def update_normal(self, mappable):
+            # [ColorizingArtist.changed() emits self.callbacks.process('changed')]
+            # Also, there is no test where self.mappable == mappable is not True
+            # and possibly no use case.
+            # Therefore, the mappable keyword can be depreciated if cm.ScalarMappable
```
```suggestion
            # Therefore, the mappable keyword can be deprecated if cm.ScalarMappable
```

QuLogic commented on lib/matplotlib/colorizer.pyi:
```diff
@@ -0,0 +1,105 @@
+from numpy.typing import ArrayLike
+
+
+class Colorizer():
```
```suggestion
class Colorizer:
```

QuLogic commented on lib/matplotlib/figure.py:
```diff
@@ -2986,7 +2986,8 @@ def set_canvas(self, canvas):
 
     @_docstring.interpd
     def figimage(self, X, xo=0, yo=0, alpha=None, norm=None, cmap=None,
-                 vmin=None, vmax=None, origin=None, resize=False, **kwargs):
+                 vmin=None, vmax=None, colorizer=None, origin=None, resize=False,
+                 **kwargs):
```
This one isn't keyword-only yet.

trygvrad replied:
I added a `*`
```py
    def figimage(self, X, xo=0, yo=0, alpha=None, norm=None, cmap=None,
                 vmin=None, vmax=None, origin=None, resize=False, *,
                 colorizer=None, **kwargs):
```
also to the typing figure.pyi, now reads:
```py
        resize: bool = ...,
        *,
        colorizer: Colorizer | None = ...,
        **kwargs
    ) -> FigureImage: ...
```

QuLogic commented on lib/matplotlib/meson.build:
```diff
@@ -31,6 +31,7 @@ python_sources = [
   'cm.py',
   'collections.py',
   'colorbar.py',
+  'colorizer.py',
```
`colorizer.pyi` needs to go in `typing_sources` as well.

tacaswell commented:
The PR cleanliness is annoying, but can be dealt with by squashing some commit together.

I'm more worried about the docs failures, not sure if the force-push that just happened was to rebase on main or not (which might pick up the updates to the skip list), but I suspect the problem is that somehow back-compat imports are being missed by automodule?

trygvrad commented:
The rebase was to main.

Not sure how I can deal with the docs issue, but let me know if there is anything I can do to troubleshoot.
Building the docs gives me
```py

...

reading sources... [100%] users/resources/index
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/trygvrad/matplotlib/matplotlib/doc/_tags/.rst: WARNING: document isn't included in any toctree
done
preparing documents... done

...

making refresh html file: users/whats_new_old redirect to users/release_notes
making refresh html file: resources/index redirect to users/resources/index
build finished with problems, 1 warning.
make: *** [Makefile:50: html] Error 1
```

tacaswell commented:
I had theories on how to fix the docs, but they all fail in interesting ways.

Writing out the docs of `cm.ScalarMappable` entirely "by hand" may be the best path.

tacaswell commented:
So things I tried:

 - adding `:imported-members:` -> fails because it _does_ find `cm.ScalarMappable`, sphinx detects that it is an alias for another class (because we changed the name) and only documents it as an attribute and thus won't include any of the methods (no matter what you do!).  This is unconditional behavior: https://github.com/sphinx-doc/sphinx/blob/ce0d1791d680e31eab634195768691fc947321af/sphinx/ext/autodoc/__init__.py#L1487-L1488
 - manually adding `.. autoclass::` to the `cm_api.rst` files fails the same way
 - try monkey-patching `ScalarMappable.__name__` in conf.py (per https://stackoverflow.com/questions/38765577/overriding-sphinx-autodoc-alias-of-for-import-of-private-class#comment115374850_58982001) but that fails because now the version in `matplotlib.colorizer` decides it is the alias and all of the `SubClass.method` links fail because they are generated by to point at the version in `colorizer` which now is not generated
 - making the class public in both modules fails because now `ScalarMappable.method` is ambiguous and a bunch of stuff would need to be re-pointed

tacaswell commented:
I think I fixed the docs, but the `ScalarMappable` API docs probably needs to be more fully filled out with the methods it used to have.

The trick to getting a sphinx-formatted docstring is:

```python
import numpydoc.docscrape_sphinx
import matplotlib.colorizer
print(str(numpydoc.docscrape_sphinx.SphinxDocString(matplotlib.colorizer._ScalarMappable.changed.__doc__)))
```

QuLogic commented on lib/matplotlib/contour.py:
```diff
@@ -1731,10 +1743,10 @@ def _initialize_x_y(self, z):
         properties of its colormap are changed. Therefore, an explicit
-        call `.QuadContourSet.changed()` is needed after modifying the
+        call `~.ContourSet.changed()` is needed after modifying the
         colormap. The explicit call can be left out, if a colorbar is
         assigned to the `.QuadContourSet` because it internally calls
-        `.QuadContourSet.changed()`.
+        `~.ContourSet.changed()`.
```
There is an exception for `QuadContourSet.changed()` in `missing-references.json` that can be deleted due to this change?

tacaswell replied:
yes.  It may also make sense to drop this change in this PR and just update the exceptions.

I also went down a rabbit hole of sorting out how we generate these tables and why the other failures happen.

QuLogic commented:
> I think I fixed the docs, but the `ScalarMappable` API docs probably needs to be more fully filled out with the methods it used to have.

How much do we want documented in `cm.ScalarMappable` vs the base `_ColorizerInterface`? Because doing something like:
```patch
diff --git a/doc/conf.py b/doc/conf.py
index d153aead73..e6cd59b64a 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -849,3 +849,7 @@ def setup(app):
     app.add_config_value('releaselevel', bld_type, 'env')
     if sphinx.version_info[:2] < (7, 1):
         app.connect('html-page-context', add_html_cache_busting, priority=1000)
+    import matplotlib.cm
+    matplotlib.cm.ScalarMappable.__name__ = 'ScalarMappable'
+    matplotlib.cm.ScalarMappable.__module__ = 'matplotlib.cm'
+    matplotlib.cm.ScalarMappable.__qualname__ = 'matplotlib.cm.ScalarMappable'
```
does work for me, but only `changed, `get_array`, and `set_array` are documented. However, it points to `_ColorizerInterface`, which does document the rest. On the other hand, I don't seem to be getting any of the broken reference warnings, so I'm not sure how much I trust that build.

tacaswell commented:
> I only set the __name__, not everything, I'll try with all three.

This generates 50-60 warnings for missing links to `matplotlib.colorizer._ScalarMappable.set_array` and `matplotlib.cm.ScalarMappable.set_clim`.  The first are from our generated kwdoc tables and the second are due to the generated class documentation not including inhereted methods so `set_clim` and `set_cmap` are not documented at all!

tacaswell commented:
On the call the consensus was to put a code snippet in `conf.py` to use numpydoc to scrape the docsting and generate a static rst file to be included so they stay in sync.

tacaswell commented:
I took the liberty of force-pushing to:

 1. re-write history to get rid of the API change file
 2. drop unneeded reference exceptions

tacaswell commented:
3. to merge the two commits that touch a test image into one.

QuLogic commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+        self.norm = norm
+
+        self.callbacks = cbook.CallbackRegistry(signals=["changed"])
+        self.colorbar = None
```
There used to be some documentation on `colorbar`.

QuLogic commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+        # bury import to avoid circular imports
+        from matplotlib import cm
+        in_init = self._cmap is None
+        cmap = cm._ensure_cmap(cmap)
+        self._cmap = cmap
```
Should fit together like it used to.
```suggestion
        self._cmap = cm._ensure_cmap(cmap)
```

QuLogic commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+
+    @property
+    def vmin(self):
+        return self.get_clim[0]
```
This doesn't look right.

trygvrad replied:
Thanks for spotting this.

I updated this and added a test to the end of colors.py, more on this further down.

QuLogic commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+
+    @property
+    def vmax(self):
+        return self.get_clim[1]
```
Also doesn't seem right.

trygvrad commented:
Thank you @QuLogic, I'll take a look at this tomorrow. 
PS: I won't make the meeting today.

trygvrad commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+)
+
+
+class ColorizingArtist(_ScalarMappable, artist.Artist):
```
I added:
```py
    """
    Base class for artists that make map data to color using a
    `.colorizer.Colorizer`.

    The `.colorizer.Colorizer` applies data normalization before
    returning RGBA colors from a `~matplotlib.colors.Colormap`,
    `~matplotlib.colors.BivarColormap` or `~matplotlib.colors.MultivarColormap`.

    """
```

trygvrad commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+    def _set_colorizer_check_keywords(self, colorizer, **kwargs):
+        """
+        Raises a ValueError if any kwarg is not None while colorizer is not None
+        Passes or creates a Colorizer object.
```
Well spotted, I have updated the docstring to reflect this

trygvrad commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+    """
+    A mixin class to map one or multiple sets of scalar data to RGBA.
+
+    The ScalarMappable applies data normalization before returning RGBA colors
+    from the given `~matplotlib.colors.Colormap`, or
+    `~matplotlib.colors.BivarColormap`.
```
I updated this to:

```
    The ScalarMappable applies data normalization before returning RGBA colors from
    the given `~matplotlib.colors.Colormap`, `~matplotlib.colors.BivarColormap`, or
    `~matplotlib.colors.MultivarColormap`.
```
It may be a bit premature to include BivarColormap and MultivarColormap in this docstring since the MultiNorm is not in yet, but I have the branch ready, just need this to be approved before I make the PR.

QuLogic replied:
You can't add this yet as it breaks the docs.

trygvrad commented:
@QuLogic Thanks again for the feedback.

I added a test `test_colorizer_vmin_vmax()` to the bottom of `test_colors.py` because @QuLogic pointed out an error in this and once I fixed it I wanted to make sure there is a test for it. That said, `test_colors.py` is not the right place for it, and I think we should really make a `test_colorizer.py`. Also given the nature of the error @QuLogic found, I think we should work to add more tests to the changes introduced in this PR. 

However I would really prefer that this PR gets approved and merged as soon as possible, then I can make the next PR (a `MultiNorm` class and exposing `BivarColormap` and `MultivarColormap` in the top level functions). That way we can get the ball rolling on this while at the same time making a separate PR that makes `test_colorizer.py` and adds additional testing to the `Colorizer` and `ColorizingArtist` classes.

I.e. I would prefer that this is approved now, and then I make an issue to add more testing (I can make the PR as well), rather than postponing this PR further until additional tests are made. 

---

> There used to be a docstring on colorbar.

My apologies for my confusion on this, @QuLogic do you want a docstring on the `@property def colorbar()` or where should this docstring be?

QuLogic commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,692 @@
+
+class ColorizingArtist(_ScalarMappable, artist.Artist):
+    """
+    Base class for artists that make map data to color using a
+    `.colorizer.Colorizer`.
```
```suggestion
    Base class for artists that map data to color using a `.colorizer.Colorizer`.
```

QuLogic commented on lib/matplotlib/colorizer.py:
```diff
@@ -0,0 +1,682 @@
+        """
+        self._colorizer.autoscale_None(self._A)
+
+    @property
+    def colorbar(self):
+        return self._colorizer.colorbar
+
+    @colorbar.setter
+    def colorbar(self, colorbar):
+        self._colorizer.colorbar = colorbar
```
Yes, the `colorbar` docstring could go on the property.

trygvrad commented:
@QuLogic These should be fixed now, I just amended the previous commit.

---

> Yes, the colorbar docstring could go on the property.

It now reads:
```py

    @property
    def colorbar(self):
        """
        The last colorbar associated with this object. May be None
        """
        return self._colorizer.colorbar
```

trygvrad commented:
@timhoffm did you find the time to look at this?

@tacaswell Its asking for a rebase because  https://github.com/matplotlib/matplotlib/pull/28658/commits/7e2c3f11b15b98a0f68ea88613f558165a5b4e5b collides with https://github.com/matplotlib/matplotlib/commit/5db8071e7647733107274154a730e196c0ab2c3b 
```
<<<<<<< HEAD
        call ``QuadContourSet.changed()`` is needed after modifying the
        colormap. The explicit call can be left out, if a colorbar is
        assigned to the `.QuadContourSet` because it internally calls
        ``QuadContourSet.changed()``.
=======
        call `~.ContourSet.changed()` is needed after modifying the
        colormap. The explicit call can be left out, if a colorbar is
        assigned to the `.QuadContourSet` because it internally calls
        `~.ContourSet.changed()`.
>>>>>>> 7e2c3f11b1 (DOC: fix unrelated xref issues)
```
i.e. should there be 
```
``QuadContourSet.changed()``
```
or 
```
`~.ContourSet.changed()`
```
or perhaps even some other combination in the docs, and I don't really know enough about the docs to resolve this.

tacaswell commented:
should be `~.ContourSet.changed()` for ... reasons.

tacaswell commented:
I am going to make the executive decision to merge this as-is (after the rebase).

trygvrad commented:
> I am going to make the executive decision to merge this as-is (after the rebase).

Thank you :)
and thank you @ksunden for rebasing this. 

Once this is merged, I'll try to get the next round of PRs going, so we can get those in for the next release (April?).

QuLogic commented:
The doc failure appears relevant, unfortunately.

tacaswell commented:
Thank you @trygvrad !  Congrats on driving a pretty big change to Matplotlib
